File: sdk/python/symbolchain/core/symbol/IdGenerator.py
import sha3

NAMESPACE_FLAG = 1 << 63


def generate_mosaic_id(owner_address, nonce):
    """Generates a mosaic id from an owner address and a nonce."""
    hasher = sha3.sha3_256()
    hasher.update(nonce.to_bytes(4, 'little'))
    hasher.update(owner_address.bytes)
    digest = hasher.digest()

    result = int.from_bytes(digest[0:8], 'little')
    if result & NAMESPACE_FLAG:
        result -= NAMESPACE_FLAG

    return result


def generate_namespace_id(name, parent_namespace_id=0):
    """Generates a namespace id from a name and an optional parent namespace id."""
    hasher = sha3.sha3_256()
    hasher.update(parent_namespace_id.to_bytes(8, 'little'))
    hasher.update(name.encode('utf8'))
    digest = hasher.digest()

    result = int.from_bytes(digest[0:8], 'little')
    return result | NAMESPACE_FLAG


def generate_mosaic_alias_id(fully_qualified_name):
    """Generates a mosaic id from a fully qualified mosaic alias name."""
    return generate_namespace_path(fully_qualified_name)[-1]


def is_valid_namespace_name(name):
    """Returns true if a name is a valid namespace name."""
    def is_alphanum(character):
        return 'a' <= character <= 'z' or '0' <= character <= '9'

    return name and is_alphanum(name[0]) and all(is_alphanum(ch) or ch in ['_', '-'] for ch in name)


def generate_namespace_path(fully_qualified_name):
    """Parses a fully qualified namespace name into a path."""
    path = []
    parent_namespace_id = 0
    for name in fully_qualified_name.split('.'):
        if not is_valid_namespace_name(name):
            raise ValueError('fully qualified name is invalid due to invalid part name ({})'.format(fully_qualified_name))

        path.append(generate_namespace_id(name, parent_namespace_id))
        parent_namespace_id = path[-1]

    return path

File: sdk/javascript/src/symbol/idGenerator.js
/* eslint-disable no-unused-vars */
import { Address } from './Network.js';
/* eslint-enable no-unused-vars */
import { sha3_256 } from '@noble/hashes/sha3';

const NAMESPACE_FLAG = 1n << 63n;

const uint32ToBytes = value => new Uint8Array([
	value & 0xFF,
	(value >> 8) & 0xFF,
	(value >> 16) & 0xFF,
	(value >> 24) & 0xFF
]);

const digestToBigInt = digest => {
	let result = 0n;
	for (let i = 0; 8 > i; ++i)
		result += (BigInt(digest[i]) << BigInt(8 * i));

	return result;
};

/**
 * Generates a mosaic id from an owner address and a nonce.
 * @param {Address} ownerAddress Owner address.
 * @param {number} nonce Nonce.
 * @returns {bigint} Computed mosaic id.
 */
const generateMosaicId = (ownerAddress, nonce) => {
	const hasher = sha3_256.create();
	hasher.update(uint32ToBytes(nonce));
	hasher.update(ownerAddress.bytes);
	const digest = hasher.digest();

	let result = digestToBigInt(digest);
	if (result & NAMESPACE_FLAG)
		result -= NAMESPACE_FLAG;

	return result;
};

/**
 * Generates a namespace id from a name and an optional parent namespace id.
 * @param {string} name Namespace name.
 * @param {bigint} parentNamespaceId Parent namespace id.
 * @returns {bigint} Computed namespace id.
 */
const generateNamespaceId = (name, parentNamespaceId = 0n) => {
	const hasher = sha3_256.create();
	hasher.update(uint32ToBytes(Number(parentNamespaceId & 0xFFFFFFFFn)));
	hasher.update(uint32ToBytes(Number((parentNamespaceId >> 32n) & 0xFFFFFFFFn)));
	hasher.update(name);
	const digest = new Uint8Array(hasher.digest());

	const result = digestToBigInt(digest);
	return result | NAMESPACE_FLAG;
};

/**
 * Returns true if a name is a valid namespace name.
 * @param {string} name Namespace name to check.
 * @returns {boolean} true if the specified name is valid.
 */
const isValidNamespaceName = name => {
	const isAlphanum = character => ('a' <= character && 'z' >= character) || ('0' <= character && '9' >= character);
	if (!name || !isAlphanum(name[0]))
		return false;

	for (let i = 0; i < name.length; ++i) {
		const ch = name[i];
		if (!isAlphanum(ch) && '_' !== ch && '-' !== ch)
			return false;
	}

	return true;
};

/**
 * Parses a fully qualified namespace name into a path.
 * @param {string} fullyQualifiedName Fully qualified namespace name.
 * @returns {Array<bigint>} Computed namespace path.
 */
const generateNamespacePath = fullyQualifiedName => {
	const path = [];
	let parentNamespaceId = 0n;
	fullyQualifiedName.split('.').forEach(name => {
		if (!isValidNamespaceName(name))
			throw Error(`fully qualified name is invalid due to invalid part name (${fullyQualifiedName})`);

		path.push(generateNamespaceId(name, parentNamespaceId));
		parentNamespaceId = path[path.length - 1];
	});

	return path;
};

/**
 * Generates a mosaic id from a fully qualified mosaic alias name.
 * @param {string} fullyQualifiedName Fully qualified mosaic name.
 * @returns {bigint} Computed mosaic id.
 */
const generateMosaicAliasId = fullyQualifiedName => {
	const path = generateNamespacePath(fullyQualifiedName);
	return path[path.length - 1];
};

export {
	generateMosaicId,
	generateNamespaceId,
	isValidNamespaceName,
	generateNamespacePath,
	generateMosaicAliasId
};
